hardware_interface/src/resource_manager.cpp

```diff
@@ -51,116 +81,416 @@ class ResourceStorage
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool cleanup_hardware(HardwareT & hardware)
+  {
+    bool result = trigger_hardware_state_transition(
+      std::bind(&HardwareT::cleanup, &hardware), "cleanup", hardware.get_name(),
+      lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED);
+
+    if (result)
+    {
+      // TODO(destogl): see comment in "configure_hardware"
+      // remove all command interfaces from available list
+      for (const auto & interface : hardware_info_map_[hardware.get_name()].command_interfaces)
+      {
+        auto found_it = std::find(
+          available_command_interfaces_.begin(), available_command_interfaces_.end(), interface);
+
+        if (found_it != available_command_interfaces_.end())
+        {
+          available_command_interfaces_.erase(found_it);
+          // TODO(destogl): change output to debug
+          RCUTILS_LOG_INFO_NAMED(
+            "resource_manager", "(hardware '%s'): '%s' command removed from available list",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+        else
+        {
+          // TODO(destogl): do here error management if interfaces are only partially added into
+          // "available" list - this should never be the case!
+          RCUTILS_LOG_WARN_NAMED(
+            "resource_manager",
+            "(hardware '%s'): '%s' command interface not in available list."
+            " This can happen due to multiple calls to 'cleanup'",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+      // remove all state interfaces from available list
+      for (const auto & interface : hardware_info_map_[hardware.get_name()].state_interfaces)
+      {
+        auto found_it = std::find(
+          available_state_interfaces_.begin(), available_state_interfaces_.end(), interface);
+
+        if (found_it != available_state_interfaces_.end())
+        {
+          available_state_interfaces_.erase(found_it);
+          // TODO(destogl): change output to debug
+          RCUTILS_LOG_INFO_NAMED(
+            "resource_manager", "(hardware '%s'): '%s' state interface removed from available list",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+        else
+        {
+          // TODO(destogl): do here error management if interfaces are only partially added into
+          // "available" list - this should never be the case!
+          RCUTILS_LOG_WARN_NAMED(
+            "resource_manager",
+            "(hardware '%s'): '%s' state interface not in available list. "
+            "This can happen due to multiple calls to 'cleanup'",
+            hardware.get_name().c_str(), interface.c_str());
+        }
+      }
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool shutdown_hardware(HardwareT & hardware)
+  {
+    bool result = trigger_hardware_state_transition(
```

Comment:
I know no triggering happens here, is this part just for adding API?

Reply:
This part abstracts the “transition” calls to actual hardware and does check that the transition finished successfully (together with print output)